R/readSource.R

```diff
@@ -96,37 +96,62 @@ readSource <- function(type, subtype = NULL, subset = NULL, convert = TRUE) { #
 
     x <- .getFromCache(prefix, type, args, subtype, subset)
     if (!is.null(x)) {
+      # cache hit, return
       return(x)
     }
+    # cache miss, read from source
 
-    # cache miss, read from source file
-    if (prefix == "correct") {
-      x <- .getData(type, subtype, subset, args, "read")
-    } else if (prefix == "convert") {
-      if (type %in% getSources(type = "correct")) {
-        x <- .getData(type, subtype, subset, args, "correct")
+    if (prefix != "read") {
+      if (prefix == "convert" && type %in% getSources(type = "correct")) {
+        upstreamPrefix <- "correct"
       } else {
-        x <- .getData(type, subtype, subset, args, "read")
+        upstreamPrefix <- "read"
       }
+      # this x is passed if prefix is correct or convert
+      x <- .getData(type, subtype, subset, args, upstreamPrefix)
     }
 
-    with_dir(sourcefolder, {
+    withr::with_dir(sourcefolder, {
       ignore <- c("subtype", "subset")[c(is.null(subtype), is.null(subset))]
       if (length(ignore) == 0) ignore <- NULL
       functionname <- prepFunctionName(type = type, prefix = prefix, ignore = ignore)
       setWrapperActive("wrapperChecks")
+
+      # run the actual read/correct/convert function
+      # the locally defined x is passed if prefix is correct or convert
       x <- withMadratLogging(eval(parse(text = functionname)))
+
       setWrapperInactive("wrapperChecks")
     })
 
-    if (!is.magpie(x)) {
-      stop('Output of function "', functionname, '" is not a MAgPIE object!')
-    }
-    if (prefix == "convert") {
-      .testISO(getItems(x, dim = 1.1), functionname = functionname)
+    # ensure we are always working with a list with entries "x" and "class"
+    xList <- if (is.list(x)) x else list(x = x)
+
+    # return list is only available for read, not for correct/convert
```

Comment:
why is the return list only available for read? it should be also available in the other two cases I think